gobblin-core/src/main/java/gobblin/converter/EnvelopeSchemaConverter.java

```diff
@@ -0,0 +1,86 @@
+import java.nio.ByteBuffer;
+import javax.xml.bind.DatatypeConverter;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+/**
+ * A converter for opening GoldenGate envelope schema.
+ */
+public class EnvelopeSchemaConverter extends AvroToAvroConverterBase {
+
+  public static final String PAYLOAD_SCHEMA_ID = "envelope.schema.id.field";
+  public static final String PAYLOAD = "envelope.payload.field";
+
+  private Optional<AvroSchemaFieldRemover> fieldRemover;
+  private KafkaAvroSchemaRegistry registry;
+
+  /**
+   * To remove certain fields from the Avro schema or records of a topic/table, set property
+   * {topic/table name}.remove.fields={comma-separated, fully qualified field names} in workUnit.
+   *
+   * E.g., PageViewEvent.remove.fields=header.memberId,mobileHeader.osVersion
+   */
+  @Override
+  public EnvelopeSchemaConverter init(WorkUnitState workUnit) {
+    if (workUnit.contains(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY)) {
+      String removeFieldsPropName = workUnit.getProp(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY) + AvroProjectionConverter.REMOVE_FIELDS;
+      if (workUnit.contains(removeFieldsPropName)) {
+        this.fieldRemover = Optional.of(new AvroSchemaFieldRemover(workUnit.getProp(removeFieldsPropName)));
+      } else {
+        this.fieldRemover = Optional.absent();
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Do nothing, actual schema must be obtained from records.
+   */
+  @Override
+  public Schema convertSchema(Schema inputSchema, WorkUnitState workUnit) throws SchemaConversionException {
+    return inputSchema;
+  }
+
+  /**
+   * Get actual schema from registry and deserialize payload using it.
+   */
+  @Override
+  public Iterable<GenericRecord> convertRecord(Schema outputSchema, GenericRecord inputRecord, WorkUnitState workUnit)
+      throws DataConversionException {
+    try {
+      if (registry == null) {
+        registry = new KafkaAvroSchemaRegistry(workUnit.getProperties());
+      }
+      String schemaKey = String.valueOf(inputRecord.get(workUnit.getProp(PAYLOAD_SCHEMA_ID)));
+      Schema payloadSchema = registry.getSchemaByKey(schemaKey);
+      byte[] payload = getPayload(inputRecord, workUnit.getProp(PAYLOAD));
+      GenericRecord outputRecord = AvroUtils.slowDeserializeGenericRecord(payload, payloadSchema);
+      if (this.fieldRemover.isPresent()) {
+        payloadSchema = this.fieldRemover.get().removeFields(payloadSchema);
+      }
+      return new SingleRecordIterable<>(AvroUtils.convertRecordSchema(outputRecord, payloadSchema));
+    } catch (IOException | SchemaRegistryException e) {
+      throw new DataConversionException(e);
+    }
+  }
+
+  /**
+   * Get payload field from GenericRecord and convert to byte array
+   */
+  public byte[] getPayload(GenericRecord inputRecord, String payloadFieldName) {
+    ByteBuffer bb = (ByteBuffer) inputRecord.get(payloadFieldName);
+    String hexString = new String(bb.array());
```

Comment:
Again, using bb.array() is unsafe. The correct code is something along the lines of 

    byte[] payloadBytes;
    if (bb.hasArray()) {
        payloadBytes = bb.array();
    } else {
       payloadBytes = new byte[bb.remaining()];
       bb.get(payloadBytes);
    }
    String hexString = new String(payloadBytes, Charset.UTF_8);

If we avoid the copying of bytes from payloadBytes to hexString to the decoded result array, it will be even better. Writing a simple function read from the ByteBuffer directly and decode on-the-fly should not be that hard.